airflow/models/taskinstance.py

```diff
@@ -992,7 +1004,11 @@ def get_triggering_events() -> dict[str, list[AssetEvent | AssetEventPydantic]]:
         # Re-attach it if we get called.
         nonlocal dag_run
         if dag_run not in session:
-            dag_run = session.merge(dag_run, load=False)
+            # In case, refresh_from_db has also included the dag_run,
+            # the object will be considered dirty by the session.
+            # Trying to merge the dirty dag_run with load=False, will result to an SQLAlchemy error.
```

Comment:
What was the error? Ideally we shouldn't need to change this

Reply:
@ashb I haven't been able to reproduce the error but it's related to 
```
_set_ti_attrs(task_instance, ti, include_dag_run=include_dag_run)
```

The issue is that we need the ti to pickup any changes to the dag_run context_carrier.

For example, the scheduler that has been running a dag is no longer available and another one picks it up. In that case, the new scheduler has to create a new span for the dag_run and needs to update the context_carrier to point to the info of the new span. Any tasks running from that point until the end, will be children of the new span. For that to work, the ti must reference the new span carrier and not the old one.

It's not as simple as always including the dag_run because there are cases where it's detached or not available and we can't retrieve it from the db.

Reply:
I removed it. It was an issue with how `include_dag_run` was getting evaluated that led me to this change.

I'm resolving this conversation.

Reply:
After finishing with all the changes, now I can see the error on the PR CI

https://github.com/apache/airflow/actions/runs/12394228248/job/34597663248?pr=43941#step:7:5611

```
sqlalchemy.exc.InvalidRequestError: merge() with load=False option
does not support objects marked as 'dirty'.
flush() all changes on mapped instances before merging with load=False.
```

Reply:
The error wasn't actually related to this. The root cause was the `include_dag_run`.